pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/FilterOperatorTest.java

```diff
@@ -0,0 +1,212 @@
+    Assert.assertEquals(result.get(1)[0], 1);
+  }
+
+  @Test
+  public void testFalseBooleanFilter() {
+    RexExpression booleanLiteral = new RexExpression.Literal(FieldSpec.DataType.BOOLEAN, false);
+
+    DataSchema inputSchema = new DataSchema(new String[]{"intCol"}, new DataSchema.ColumnDataType[]{
+        DataSchema.ColumnDataType.INT
+    });
+    Mockito.when(_upstreamOperator.nextBlock())
+        .thenReturn(OperatorTestUtil.block(inputSchema, new Object[]{1}, new Object[]{2}));
+    FilterOperator op = new FilterOperator(_upstreamOperator, inputSchema, booleanLiteral);
+    TransferableBlock dataBlock = op.getNextBlock();
+    Assert.assertFalse(dataBlock.isErrorBlock());
+    List<Object[]> result = dataBlock.getContainer();
+    Assert.assertTrue(result.isEmpty());
+  }
+
+  @Test(expectedExceptions = IllegalStateException.class, expectedExceptionsMessageRegExp = ".*boolean literal.*")
+  public void testBadTypeLiteralFilter() {
+    RexExpression booleanLiteral = new RexExpression.Literal(FieldSpec.DataType.STRING, "false");
+    DataSchema inputSchema = new DataSchema(new String[]{"intCol"}, new DataSchema.ColumnDataType[]{
+        DataSchema.ColumnDataType.INT
+    });
+    Mockito.when(_upstreamOperator.nextBlock())
+        .thenReturn(OperatorTestUtil.block(inputSchema, new Object[]{1}, new Object[]{2}));
+    FilterOperator op = new FilterOperator(_upstreamOperator, inputSchema, booleanLiteral);
+  }
+
+  @Test(expectedExceptions = IllegalStateException.class, expectedExceptionsMessageRegExp = ".*Input has to be "
+      + "boolean type.*")
+  public void testInputRefFilterBadType() {
+    RexExpression ref0 = new RexExpression.InputRef(0);
+    DataSchema inputSchema = new DataSchema(new String[]{"intCol"}, new DataSchema.ColumnDataType[]{
+        DataSchema.ColumnDataType.INT
+    });
+    Mockito.when(_upstreamOperator.nextBlock())
+        .thenReturn(OperatorTestUtil.block(inputSchema, new Object[]{1}, new Object[]{2}));
+    FilterOperator op = new FilterOperator(_upstreamOperator, inputSchema, ref0);
+  }
+
+  @Test
+  public void testInputRefFilter() {
+    RexExpression ref1 = new RexExpression.InputRef(1);
+    DataSchema inputSchema = new DataSchema(new String[]{"intCol", "boolCol"}, new DataSchema.ColumnDataType[]{
+        DataSchema.ColumnDataType.INT, DataSchema.ColumnDataType.BOOLEAN
+    });
+    Mockito.when(_upstreamOperator.nextBlock())
+        .thenReturn(OperatorTestUtil.block(inputSchema, new Object[]{1, true}, new Object[]{2, false}));
+    FilterOperator op = new FilterOperator(_upstreamOperator, inputSchema, ref1);
+    TransferableBlock dataBlock = op.getNextBlock();
+    Assert.assertFalse(dataBlock.isErrorBlock());
+    List<Object[]> result = dataBlock.getContainer();
+    Assert.assertEquals(result.size(), 1);
+    Assert.assertEquals(result.get(0)[0], 1);
+    Assert.assertEquals(result.get(0)[1], true);
+  }
+
+  @Test
+  public void testAndFilter() {
```

Comment:
suggestion: instead of testing the entire `FilterOperator`, why not make each of the operand classes (e.g. `AndOperand`) package private so that we can test only that specific code? (I believe that's what you are testing here - and that also makes sure that anywhere it's reused also has coverage for it, such as `HashJoinOperator`)

Reply:
This is because we don't have time to write operand unit test. 
Another thing is we should test via API instead of test implementation.  I feel FilterOperator is our end API. 
It could be FilterOperand implementation is correct but we use it wrong in FilterOperator. 

Think about the case for MailboxOperators... since we mock underlying service, we are not able to catch the errors. 
Let me know if you still disagree. I am open to discuss this further.

Reply:
for the sake of argument (I'm happy to leave it as is) `I feel FilterOperator is our end API` is where I think we disagree 😉 

I feel that "our end API" is SQL. `FilterOperator` has only one job/API: take rows and pass them to the underlying operator. If it does that, then it did its job correctly.  Similarly, each operand should do its job correctly - the API for an operand is to take a row and determine whether or not that row should be included.

To make sure they work together when a SQL query is issued is an integration test, and that should be tested at the top level (via the new JSON query runner tests).

Reply:
Yea, I agree SQL is our real API. I would say I agree with you if we have the extra time to add each operand test. but let's leave it as it is now and slowly add operand unit test.

Reply:
definitely, let's not block on this! thanks for the discussion :)